src/middleware/operation.rs

```diff
@@ -370,79 +370,87 @@ pub fn check_st_tmpl(
     st_arg: &StatementArg,
     // Map from wildcards to values that we have seen so far.
     wildcard_map: &mut [Option<Value>],
-) -> bool {
+) -> Result<()> {
     // Check that the value `v` at wildcard `wc` exists in the map or set it.
-    fn check_or_set(v: Value, wc: &Wildcard, wildcard_map: &mut [Option<Value>]) -> bool {
+    fn check_or_set(v: Value, wc: &Wildcard, wildcard_map: &mut [Option<Value>]) -> Result<()> {
         if let Some(prev) = &wildcard_map[wc.index] {
             if *prev != v {
-                // TODO: Return nice error
-                return false;
+                return Err(Error::invalid_wildcard_assignment(
+                    wc.clone(),
+                    v,
+                    prev.clone(),
+                ));
             }
         } else {
             wildcard_map[wc.index] = Some(v);
         }
-        true
+        Ok(())
     }
 
     match (st_tmpl_arg, st_arg) {
-        (StatementTmplArg::None, StatementArg::None) => true,
-        (StatementTmplArg::Literal(lhs), StatementArg::Literal(rhs)) if lhs == rhs => true,
+        (StatementTmplArg::None, StatementArg::None) => Ok(()),
+        (StatementTmplArg::Literal(lhs), StatementArg::Literal(rhs)) if lhs == rhs => Ok(()),
         (
             StatementTmplArg::AnchoredKey(pod_id_wc, key_tmpl),
             StatementArg::Key(AnchoredKey { pod_id, key }),
         ) => {
             let pod_id_ok = check_or_set(Value::from(*pod_id), pod_id_wc, wildcard_map);
-            pod_id_ok && (key_tmpl == key)
+            pod_id_ok.and_then(|_| {
+                (key_tmpl == key).then_some(()).ok_or(
+                    Error::mismatched_anchored_key_in_statement_tmpl_arg(
+                        key_tmpl.clone(),
+                        key.clone(),
+                    ),
```

Comment:
This error only has context for the check `key_tmpl == key`.
But it will also be returned if `pod_id_ok = false`.

I have two alternative suggestions:
a. Extend the error to also include the context of the pod_id
b. Add a new error just for `pod_id_ok = false`.